src/mono/browser/runtime/http.ts

```diff
@@ -66,167 +81,188 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
+    if (BuildConfiguration === "Debug") commonAsserts(controller);
     // the bodyPtr is pinned by the caller
     const view = new Span(bodyPtr, bodyLength, MemoryViewType.Byte);
     const copy = view.slice() as Uint8Array;
-    return http_wasm_fetch(url, header_names, header_values, option_names, option_values, abort_controller, copy);
+    return http_wasm_fetch(controller, url, header_names, header_values, option_names, option_values, copy);
 }
 
-export function http_wasm_fetch(url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], abort_controller: AbortController, body: Uint8Array | ReadableStream | null): ControllablePromise<ResponseExtension> {
+export function http_wasm_fetch(controller: HttpController, url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], body: Uint8Array | ReadableStream | null): ControllablePromise<void> {
+    if (BuildConfiguration === "Debug") commonAsserts(controller);
     verifyEnvironment();
     assert_js_interop();
     mono_assert(url && typeof url === "string", "expected url string");
     mono_assert(header_names && header_values && Array.isArray(header_names) && Array.isArray(header_values) && header_names.length === header_values.length, "expected headerNames and headerValues arrays");
     mono_assert(option_names && option_values && Array.isArray(option_names) && Array.isArray(option_values) && option_names.length === option_values.length, "expected headerNames and headerValues arrays");
+
     const headers = new Headers();
     for (let i = 0; i < header_names.length; i++) {
         headers.append(header_names[i], header_values[i]);
     }
     const options: any = {
         body,
         headers,
-        signal: abort_controller.signal
+        signal: controller.abortController.signal
     };
     if (typeof ReadableStream !== "undefined" && body instanceof ReadableStream) {
         options.duplex = "half";
     }
     for (let i = 0; i < option_names.length; i++) {
         options[option_names[i]] = option_values[i];
     }
-
-    return wrap_as_cancelable_promise(async () => {
-        const res = await loaderHelpers.fetch_like(url, options) as ResponseExtension;
-        res.__abort_controller = abort_controller;
-        return res;
+    // make the fetch cancellable
+    controller.fetchResponsePromise = wrap_as_cancelable_promise(() => {
+        return loaderHelpers.fetch_like(url, options);
     });
-}
-
-function get_response_headers(res: ResponseExtension): void {
-    if (!res.__headerNames) {
-        res.__headerNames = [];
-        res.__headerValues = [];
+    // avoid processing headers if the fetch is cancelled
+    controller.fetchResponsePromise.then((res: Response) => {
+        controller.fetchResponse = res;
+        controller.responseHeaderNames = [];
+        controller.responseHeaderValues = [];
```

Comment:
nit: Please unify name pattern, eg `fetchResponse` and `fetchResponseHeader*` or `response` and `reponseHeader*`